lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp

```diff
@@ -1143,37 +1143,122 @@ bool SinkReshapes::run(Function *F, const CompilationContext &cctx) {
+        newN->setNthInput(0, in);
+        newN->setTypeUnsafe(0, newTy);
+        newN->setPredicate(node->getPredicate());
+
+        // Create new Reshape.
+        auto *newRS = F->createReshape(RS->getName(), newN,
+                                       RS->getResult().getType()->dims());
+        newRS->setPredicate(node->getPredicate());
+        out.replaceAllUsesOfWith(newRS->getResult());
 
-    auto *RS = dyn_cast<ReshapeNode>(node->getNthInput(0));
-    if (!RS) {
-      continue;
-    }
+        changed = true;
+      }
+
+      // Binary eltwise nodes.
+      if (node->getNumInputs() == 2 && node->getNumResults() == 1) {
+        DCHECK(node->getNthResult(0).getType()->isFusedQuantizedType() ||
+               node->getNthInput(0).getType()->isFusedQuantizedType() ||
+               node->getNthResult(0).dims().equals(node->getNthInput(0).dims()))
+            << "SinkReshapes: not an element-wise node: " << node->toString();
+        DCHECK(node->getNthResult(0).getType()->isFusedQuantizedType() ||
+               node->getNthInput(1).getType()->isFusedQuantizedType() ||
+               node->getNthResult(0).dims().equals(node->getNthInput(1).dims()))
+            << "SinkReshapes: not an element-wise node: " << node->toString();
+
+        // At least one of the inputs must be a Reshape.
+        // If both inputs are Reshapes, they must have the same dimensions.
+        auto *LRN = dyn_cast<ReshapeNode>(node->getNthInput(0));
+        auto *RRN = dyn_cast<ReshapeNode>(node->getNthInput(1));
+        if (!LRN && !RRN) {
+          continue;
+        }
+        if (LRN && RRN &&
+            !LRN->getResult().dims().equals(RRN->getResult().dims())) {
+          continue;
+        }
 
-    // Create new eltwise node.
-    auto in = RS->getInput();
-    auto out = node->getNthResult(0);
-    auto newTy =
-        F->getParent()->uniqueTypeWithNewShape(out.getType(), in.dims());
-    auto *newN = F->addNode(node->clone());
-    newN->setNthInput(0, in);
-    newN->setTypeUnsafe(0, newTy);
-    newN->setPredicate(node->getPredicate());
+        // Canonicalize node to simplify transformation implementation (make LHS
+        // always be the input with a Reshape).
+        bool swap = (LRN == nullptr);
+        auto nv = node->getNthInput(1);
+        if (swap) {
+          nv = node->getNthInput(0);
+          LRN = RRN;
+          RRN = nullptr;
+        }
 
-    // Create new Reshape.
-    auto *newRS = F->createReshape(RS->getName(), newN,
-                                   RS->getResult().getType()->dims());
-    newRS->setPredicate(node->getPredicate());
-    out.replaceAllUsesOfWith(newRS->getResult());
+        // RHS must be either a Reshape or a Constant (+ Quantize) or Splat.
+        auto *RQ = dyn_cast<QuantizeNode>(nv);
+        auto *RC = dyn_cast<Constant>(RQ ? RQ->getInput() : nv);
+        auto *RS = dyn_cast<SplatNode>(nv);
+        if (!RRN && !RC && !RS) {
+          continue;
+        }
```

Comment:
Combined with the below suggestion of not relying on specific nodes -- can we just check if this is a constant chain of operations and fold if so? I.e. we could make `isConstantOperation` from `ConstantFolding.cpp` usable outside of there and call it here, e.g.

```suggestion
        // RHS must be a constant chain if it's not a reshape, to allow for the reshape to get folded into the chain later on.
        if (!RRN || !isConstantOperation(nv)) {
          continue;
        }
```

(We'd need to do something like assume Interpreter is the constant folding backend, but I think that's generally the case anyway...)

Reply:
If we use `isConstantOperation`, we'll change the algorithm from O(N) to something like O(N^2). But the optimization should indeed benefit from it. Is that ok?